tree-sitter/src/TreeSitter/GenerateSyntax.hs

```diff
@@ -50,82 +50,93 @@ astDeclarationsForLanguage language filePath = do
-  where child = TH.bangType (TH.bang noSourceUnpackedness noSourceStrictness) (conT (toName named n))
+constructorForSumChoice :: String -> Name -> TreeSitter.Deserialize.Type -> Q Con
+constructorForSumChoice str typeParameterName (MkType (DatatypeName n) named) = normalC (toName named (n ++ str)) [child]
+  where child = TH.bangType (TH.bang noSourceUnpackedness noSourceStrictness) (appT (conT (toName named n)) (varT typeParameterName))
 
 -- | Build Q Constructor for product types (nodes with fields)
-ctorForProductType :: String -> Maybe Children -> [(String, Field)] -> Q Con
-ctorForProductType constructorName children fields = recC (toName Named constructorName) lists where
-  lists = fieldList ++ childList
-  fieldList = fmap (uncurry toVarBangType) fields
-  childList = toList $ fmap toVarBangTypeChild children
-  toVarBangType name (MkField required fieldTypes mult) =
-    let fieldName = mkName . addTickIfNecessary . removeUnderscore $ name
-        strictness = TH.bang noSourceUnpackedness noSourceStrictness
-        ftypes = fieldTypesToNestedEither fieldTypes
-        fieldContents = case (required, mult) of
-          (Required, Multiple) -> appT (conT ''NonEmpty) ftypes
-          (Required, Single) -> ftypes
-          (Optional, Multiple) -> appT (conT ''[]) ftypes
-          (Optional, Single) -> appT (conT ''Maybe) ftypes
-    in TH.varBangType fieldName (TH.bangType strictness fieldContents)
-  toVarBangTypeChild (MkChildren field) = toVarBangType "extra_children" field
+ctorForProductType :: String -> Name -> Maybe Children -> [(String, Field)] -> Q Con
+ctorForProductType constructorName typeParameterName children fields = ctorForTypes constructorName lists where
+  lists = annotation : fieldList ++ childList
+  annotation = ("ann", varT typeParameterName)
+  fieldList = map (fmap toType) fields
+  childList = toList $ fmap toTypeChild children
+  toType (MkField required fieldTypes mult) =
+    let ftypes = fieldTypesToNestedEither fieldTypes typeParameterName
+    in case (required, mult) of
+      (Required, Multiple) -> appT (conT ''NonEmpty) ftypes
+      (Required, Single) -> ftypes
+      (Optional, Multiple) -> appT (conT ''[]) ftypes
+      (Optional, Single) -> appT (conT ''Maybe) ftypes
+  toTypeChild (MkChildren field) = ("extra_children", toType field)
 
 
 -- | Build Q Constructor for leaf types (nodes with no fields or subtypes)
-ctorForLeafType :: Named -> DatatypeName -> Q Con
-ctorForLeafType Anonymous (DatatypeName name) = normalC (toName Anonymous name) []
-ctorForLeafType Named (DatatypeName name) = recC (toName Named name) [leafBytes] where
-  leafBytes = TH.varBangType (mkName "bytes") textValue
-  textValue = TH.bangType (TH.bang noSourceUnpackedness noSourceStrictness) (conT ''Text)
+ctorForLeafType :: Named -> DatatypeName -> Name -> Q Con
+ctorForLeafType Anonymous (DatatypeName name) typeParameterName = normalC (toName Anonymous name) [TH.bangType strictness (varT typeParameterName)]
```

Comment:
Should this build a record constructor instead of a normal constructor, such that we’d have an `ann :: a` field for anonymous leaves?

Reply:
`ctorForTypes` takes a `String`, but if it took a `Name` for the constructor’s name then we’d be able to use it for this case as well as the other one. It could equally take a `Type`, since `MkType` holds both a `DatatypeName` & `Named`, tho it’s not clear to me whether that would be more or less convenient.

Reply:
I think I can further simplify but as of now, we are now generating annotations for anonymous types (ex.`AnonymousPlus`):

```
Right (Module {ann = (Range {start = 0, end = 5},Span {spanStart = Pos {posLine = 0, posColumn = 0}, spanEnd = Pos {posLine = 0, posColumn = 5}}), statement = [Right (ExpressionStatementSimpleStatement (ExpressionStatement {ann = (Range {start = 0, end = 5},Span {spanStart = Pos {posLine = 0, posColumn = 0}, spanEnd = Pos {posLine = 0, posColumn = 5}}), extraChildren = [Left (PrimaryExpressionExpression (BinaryOperatorPrimaryExpression (BinaryOperator {ann = (Range {start = 0, end = 5},Span {spanStart = Pos {posLine = 0, posColumn = 0}, spanEnd = Pos {posLine = 0, posColumn = 5}}), operator = Right (Right (Right (Right (Left (AnonymousPlus {ann = (Range {start = 2, end = 3},Span {spanStart = Pos {posLine = 0, posColumn = 2}, spanEnd = Pos {posLine = 0, posColumn = 3}}), bytes = "+"}))))), left = IdentifierPrimaryExpression (Identifier {ann = (Range {start = 0, end = 1},Span {spanStart = Pos {posLine = 0, posColumn = 0}, spanEnd = Pos {posLine = 0, posColumn = 1}}), bytes = "a"}), right = IdentifierPrimaryExpression (Identifier {ann = (Range {start = 4, end = 5},Span {spanStart = Pos {posLine = 0, posColumn = 4}, spanEnd = Pos {posLine = 0, posColumn = 5}}), bytes = "b"})})))]}))]})
```